src/js/smtpserver.c

```diff
@@ -121,41 +452,452 @@ int init_header_class(JSContext *cx, JSObject *global) {
+	int sockfd;
+	bfd_t *client_stream;
+
+	client = JS_THIS(cx, vp);
+
+	// Get host
+	if (!JS_GetProperty(cx, JSVAL_TO_OBJECT(client), "host", &host)) {
+		return JS_FALSE;
+	}
+
+	// Get port
+	if (!JS_GetProperty(cx, JSVAL_TO_OBJECT(client), "port", &port)) {
+		return JS_FALSE;
+	}
+
+	c_host = JS_EncodeString(cx, JSVAL_TO_STRING(host));
+	c_port = JS_EncodeString(cx, JSVAL_TO_STRING(port));
+
+	sockfd = connect_to_address(c_host, c_port);
+
+	client_stream = bfd_alloc(sockfd);
+	clientStream = PRIVATE_TO_JSVAL(client_stream);
+
+	if (!JS_SetProperty(cx, JSVAL_TO_OBJECT(client), "clientStream", &clientStream)) {
+		return JS_FALSE;
+	}
+
+	free(c_host);
+	free(c_port);
+
+	return JS_TRUE;
+}
+
+static JSBool smtpClient_readResponse(JSContext *cx, unsigned argc, jsval *vp) {
+	jsval smtpClient, connection, content, response, clientStream;
+	jsval js_code, js_messages, js_disconnect;
+	JSObject *messages_obj, *global;
+
+	int sockfd, code, lines_count;
+	char buf[SMTP_COMMAND_MAX + 1], *p, sep;
+	ssize_t sz;
+	bfd_t *client_stream;
+
+	global = JS_GetGlobalForScopeChain(cx);
+	smtpClient = JS_THIS(cx, vp);
+	messages_obj = JS_NewArrayObject(cx, 0, 0);
+
+	if (!JS_GetProperty(cx, JSVAL_TO_OBJECT(smtpClient), "clientStream", &clientStream)) {
+		return JS_FALSE;
+	}
+
+	client_stream = JSVAL_TO_PRIVATE(clientStream);
+
+	sockfd = client_stream->fd;
+
+	lines_count = 0;
+	do {
+		sz = 0;
+		do {
+			if (read(sockfd, &buf[sz++], 1) < 0)
+				return JS_FALSE;
+		} while (buf[sz - 1] != '\n');
+		buf[sz] = '\0';
+
+		if (sz < 4)
+			return JS_FALSE;
+
+		sep = buf[3];
+		buf[3] = '\0';
+		code = strtol(buf, &p, 10);
+
+		if ((sep != ' ' && sep != '-') || *p != '\0')
+			return JS_FALSE;
+		if (code < 100 || code > 999)
+			return JS_FALSE;
+
+		if (buf[sz - 1] == '\n')
+			buf[--sz] = '\0';
+		if (buf[sz - 1] == '\r')
+			buf[--sz] = '\0';
+
+		//add response
+		content = STRING_TO_JSVAL(JS_InternString(cx, buf + 4));
+		if (!JS_SetElement(cx, messages_obj, lines_count++, &content)) {
+			return -1;
+		}
+	} while (sep == '-');
+
+	js_code = INT_TO_JSVAL(code);
+	js_messages = OBJECT_TO_JSVAL(messages_obj);
+	js_disconnect = JSVAL_FALSE;
+	jsval argv[] = {js_code, js_messages, js_disconnect};
+
+	JS_CallFunctionName(cx, global, "SmtpResponse",
+				3, argv, &response);
+
+	JS_SET_RVAL(cx, vp, response);
+	return JS_TRUE;
+}
+
+static JSBool smtpClient_sendCommand(JSContext *cx, unsigned argc, jsval *vp) {
+	jsval command, args, smtpClient, clientStream;
+	char *c_str;
+	int n;
+	bfd_t *client_stream;
+
+	command = JS_ARGV(cx, vp)[0];
+
+	if (argc > 1) {
+		args = JS_ARGV(cx, vp)[1];
+	} else {
+		args = JSVAL_NULL;
+	}
+
+	smtpClient = JS_THIS(cx, vp);
+
+	if (!JS_GetProperty(cx, JSVAL_TO_OBJECT(smtpClient), "clientStream", &clientStream)) {
+		return JS_FALSE;
+	}
+
+	client_stream = JSVAL_TO_PRIVATE(clientStream);
+
+	if (!JSVAL_IS_NULL(args)) {
+		command = STRING_TO_JSVAL(JS_ConcatStrings(cx, JSVAL_TO_STRING(command), JS_InternString(cx, " ")));
```

Comment:
Same remark as before: since the resulting concatenated string is not needed back in JS, it's easier to concatenate the string in C using struct string_buffer.